File: src/main/java/szewek/mcflux/tileentities/TileEntityEnergyMachine.java
package szewek.mcflux.tileentities;

import net.minecraft.block.state.IBlockState;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.network.NetworkManager;
import net.minecraft.network.play.server.SPacketUpdateTileEntity;
import net.minecraft.tileentity.TileEntity;
import net.minecraft.util.EnumFacing;
import net.minecraft.util.ITickable;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.World;
import szewek.mcflux.MCFluxResources;
import szewek.mcflux.U;
import szewek.mcflux.api.MCFluxAPI;
import szewek.mcflux.api.ex.Battery;
import szewek.mcflux.api.ex.IEnergy;
import szewek.mcflux.api.fe.FE;
import szewek.mcflux.api.fe.Flavored;
import szewek.mcflux.api.fe.FlavoredContainer;
import szewek.mcflux.api.fe.IFlavorEnergy;
import szewek.mcflux.blocks.BlockEnergyMachine;
import szewek.mcflux.blocks.BlockSided;
import szewek.mcflux.config.MCFluxConfig;
import szewek.mcflux.network.MCFluxNetwork;
import szewek.mcflux.network.msg.MsgUpdateClient;
import szewek.mcflux.network.msg.MsgUpdateServer;
import szewek.mcflux.util.TransferType;

import javax.annotation.Nonnull;
import java.util.function.IntBinaryOperator;

import static szewek.mcflux.config.MCFluxConfig.CHUNK_CHARGER_TRANS;

public class TileEntityEnergyMachine extends TileEntityWCEAware implements ITickable {
	private FlavoredContainer cnt = null;
	private boolean oddTick = true, clientUpdate = true, serverUpdate = false;
	private TransferType[] sideTransfer = new TransferType[]{TransferType.NONE, TransferType.NONE, TransferType.NONE, TransferType.NONE, TransferType.NONE, TransferType.NONE};
	private long[] sideValues = new long[]{0, 0, 0, 0, 0, 0};
	private IBlockState cachedState = MCFluxResources.SIDED.getDefaultState();
	private IntBinaryOperator module;
	private int moduleId;

	public IBlockState getCachedState() {
		return cachedState;
	}

	public void setModuleId(int id) {
		moduleId = id;
		module = getModule(id);
	}

	public int getModuleId() {
		return moduleId;
	}

	private IntBinaryOperator getModule(int i) {
		switch (i) {
			case 0: return this::moduleEnergyDistributor;
			case 1: return this::moduleChunkCharger;
			case 2: return this::moduleFlavorDistributor;
			case 3: return this::moduleChunkSprayer;
		}
		return null;
	}

	@Override
	public void setPos(@Nonnull BlockPos bp) {
		super.setPos(bp);

	}

	@Override public void onLoad() {
		if (worldObj.isRemote) {
			MCFluxNetwork.toServer(MsgUpdateClient.with(pos));
			clientUpdate = false;
		}
	}

	@Override
	public void update() {
		super.update();
		if (worldObj.isRemote && clientUpdate) {
			MCFluxNetwork.toServer(MsgUpdateClient.with(pos));
			clientUpdate = false;
		} else if (!worldObj.isRemote && serverUpdate) {
			MCFluxNetwork.toDimension(MsgUpdateServer.with(pos, sideTransfer), worldObj.provider.getDimension());
			serverUpdate = false;
		}
		if (!worldObj.isRemote && wce != null && ((moduleId < 2 && bat != null) || cnt != null)) {
			int i = oddTick ? 0 : 3, m = oddTick ? 3 : 6;
			for (int j = i; j < m; j++)
				sideValues[j] = 0;
			if (module != null)
				module.applyAsInt(i, m);
		}
		oddTick = !oddTick;
	}

	@Override
	public void readFromNBT(NBTTagCompound nbt) {
		super.readFromNBT(nbt);
		moduleId = nbt.getInteger("module");
		int[] sides = nbt.getIntArray("sides");
		if (sides.length != 6) return;
		TransferType[] tt = TransferType.values();
		for (int i = 0; i < 6; i++) {
			sideTransfer[i] = tt[sides[i]];
			cachedState = cachedState.withProperty(BlockSided.sideFromId(i), sides[i]);
		}
		module = getModule(moduleId);
		serverUpdate = true;
	}

	@Override
	@Nonnull
	public NBTTagCompound writeToNBT(NBTTagCompound nbt) {
		super.writeToNBT(nbt);
		int[] sides = new int[6];
		for (int i = 0; i < 6; i++) {
			sides[i] = sideTransfer[i].ord;
		}
		nbt.setIntArray("sides", sides);
		nbt.setInteger("module", moduleId);
		return nbt;
	}

	@Override
	public boolean shouldRefresh(World w, BlockPos pos, @Nonnull IBlockState obs, @Nonnull IBlockState nbs) {
		return obs.getBlock() != nbs.getBlock() || obs.getValue(BlockEnergyMachine.VARIANT) != nbs.getValue(BlockEnergyMachine.VARIANT);
	}

	@Override
	public SPacketUpdateTileEntity getUpdatePacket() {
		return new SPacketUpdateTileEntity(pos, getBlockMetadata(), writeToNBT(new NBTTagCompound()));
	}

	@Override
	public void onDataPacket(NetworkManager net, SPacketUpdateTileEntity pkt) {
		readFromNBT(pkt.getNbtCompound());
	}

	@Override protected boolean updateVariables() {
		if ((updateMode & 3) != 0 && pos != null) {
			cnt = worldObj != null && !worldObj.isRemote ? wce.getFlavorEnergyChunk(pos.getX(), pos.getY(), pos.getZ()) : null;
		}
		return true;
	}

	public void switchSideTransfer(EnumFacing f) {
		int s = f.getIndex();
		int v = (sideTransfer[s].ord + 1) % 3;
		sideTransfer[s] = TransferType.values()[v];
		cachedState = cachedState.withProperty(BlockSided.sideFromId(s), v);
		MCFluxNetwork.toDimension(MsgUpdateServer.with(pos, sideTransfer), worldObj.provider.getDimension());
		markDirty();
	}

	public long getTransferSide(EnumFacing f) {
		return sideValues[f.getIndex()];
	}

	public TransferType[] getAllTransferSides() {
		return sideTransfer;
	}

	public void updateTransferSides(TransferType[] tts) {
		for (int i = 0; i < 6; i++) {
			sideTransfer[i] = tts[i];
			cachedState = cachedState.withProperty(BlockSided.sideFromId(i), tts[i].ord);
		}
	}

	private int moduleEnergyDistributor(int i, int m) {
		for (; i < m; i++) {
			TransferType tt = sideTransfer[i];
			if (tt == TransferType.NONE)
				continue;
			EnumFacing f = EnumFacing.VALUES[i];
			TileEntity te = worldObj.getTileEntity(pos.offset(f, 1));
			if (te == null)
				continue;
			f = f.getOpposite();
			IEnergy ea = MCFluxAPI.getEnergySafely(te, f);
			if (ea == null)
				continue;
			switch (tt) {
				case INPUT:
					sideValues[i] = U.transferEnergy(ea, bat, MCFluxConfig.ENERGY_DIST_TRANS * 2) / 2;
					break;
				case OUTPUT:
					sideValues[i] = U.transferEnergy(bat, ea, MCFluxConfig.ENERGY_DIST_TRANS * 2) / 2;
					break;
			}
		}
		return 0;
	}

	private int moduleChunkCharger(int i, int m) {
		for (; i < m; i++) {
			TransferType tt = sideTransfer[i];
			if (tt == TransferType.NONE)
				continue;
			EnumFacing f = EnumFacing.VALUES[i];
			BlockPos bpc = pos.offset(f, 16);
			Battery ebc = wce.getEnergyChunk(bpc.getX(), bpc.getY(), bpc.getZ());
			if (ebc == null)
				continue;
			switch (tt) {
				case INPUT:
					sideValues[i] = U.transferEnergy(ebc, bat, CHUNK_CHARGER_TRANS * 2) / 2;
					break;
				case OUTPUT:
					sideValues[i] = U.transferEnergy(bat, ebc, CHUNK_CHARGER_TRANS * 2) / 2;
					break;
				default:
			}
		}
		return 0;
	}

	private int moduleFlavorDistributor(int i, int m) {
		for (; i < m; i++) {
			TransferType tt = sideTransfer[i];
			if (tt == TransferType.NONE)
				continue;
			EnumFacing f = EnumFacing.VALUES[i];
			TileEntity te = worldObj.getTileEntity(pos.offset(f, 1));
			if (te == null)
				continue;
			f = f.getOpposite();
			IFlavorEnergy fea = te.getCapability(FE.CAP_FLAVOR_ENERGY, f);
			if (fea == null)
				continue;
			Flavored[] lfl;
			switch (tt) {
				case INPUT:
					lfl = fea.allFlavorsAcceptable();
					for (Flavored fl : lfl)
						U.transferFlavorEnergy(fea, cnt, fl, CHUNK_CHARGER_TRANS * 2);
					break;
				case OUTPUT:
					lfl = fea.allFlavorsContained();
					for (Flavored fl : lfl)
						U.transferFlavorEnergy(cnt, fea, fl, CHUNK_CHARGER_TRANS * 2);
					break;
				default:
			}
		}
		return 0;
	}

	private int moduleChunkSprayer(int i, int m) {
		for (; i < m; i++) {
			TransferType tt = sideTransfer[i];
			if (tt == TransferType.NONE)
				continue;
			EnumFacing f = EnumFacing.VALUES[i];
			BlockPos bpc = pos.offset(f, 16);
			FlavoredContainer efc = wce.getFlavorEnergyChunk(bpc.getX(), bpc.getY(), bpc.getZ());
			if (efc == null)
				continue;
			Flavored[] lfl;
			switch (tt) {
				case INPUT:
					lfl = efc.allFlavorsContained();
					for (Flavored fl : lfl)
						U.transferFlavorEnergy(efc, cnt, fl, CHUNK_CHARGER_TRANS * 2);
					break;
				case OUTPUT:
					lfl = cnt.allFlavorsContained();
					for (Flavored fl : lfl)
						U.transferFlavorEnergy(cnt, efc, fl, CHUNK_CHARGER_TRANS * 2);
					break;
				default:
			}
		}
		return 0;
	}
}

File: src/main/java/szewek/mcflux/tileentities/TileEntityEnergyMachine.kt
package szewek.mcflux.tileentities

import net.minecraft.block.state.IBlockState
import net.minecraft.nbt.NBTTagCompound
import net.minecraft.network.NetworkManager
import net.minecraft.network.play.server.SPacketUpdateTileEntity
import net.minecraft.util.EnumFacing
import net.minecraft.util.ITickable
import net.minecraft.util.math.BlockPos
import net.minecraft.world.World
import szewek.fl.FLU
import szewek.fl.annotations.NamedResource
import szewek.mcflux.MCFlux
import szewek.mcflux.blocks.BlockEnergyMachine
import szewek.mcflux.blocks.BlockSided
import szewek.mcflux.config.MCFluxConfig
import szewek.mcflux.network.MCFluxNetwork
import szewek.mcflux.network.Msg
import szewek.mcflux.util.TransferType

@NamedResource("mcflux:emachine")
class TileEntityEnergyMachine : TileEntityWCEAware(), ITickable {
	private var oddTick = true
	private var clientUpdate = true
	private var serverUpdate = false
	val allTransferSides = arrayOf(TransferType.NONE, TransferType.NONE, TransferType.NONE, TransferType.NONE, TransferType.NONE, TransferType.NONE)
	private val sideValues = longArrayOf(0, 0, 0, 0, 0, 0)
	var cachedState = MCFlux.Resources.SIDED.defaultState
		private set
	var moduleId: Int = 0

	override fun onLoad() {
		if (world.isRemote) {
			MCFluxNetwork.toServer(Msg.update(pos, emptyArray()))
			clientUpdate = false
		}
	}

	public override fun updateTile() {
		if (world.isRemote && clientUpdate) {
			MCFluxNetwork.toServer(Msg.update(pos, emptyArray()))
			clientUpdate = false
		} else if (!world.isRemote && serverUpdate) {
			MCFluxNetwork.toDimension(Msg.update(pos, allTransferSides), world.provider.dimension)
			serverUpdate = false
		}
		if (!world.isRemote && bat != null) {
			val i = if (oddTick) 0 else 3
			val m = i + 3
			for (j in i until m)
				sideValues[j] = 0
			when(moduleId) {
				0 -> moduleEnergyDistributor(i, m)
				1 -> moduleChunkCharger(i, m)
			}
		}
		oddTick = !oddTick
	}

	override fun readFromNBT(nbt: NBTTagCompound) {
		super.readFromNBT(nbt)
		moduleId = nbt.getInteger("module")
		val sides = nbt.getIntArray("sides")
		if (sides.size != 6) return
		val tt = TransferType.values()
		for (i in 0..5) {
			allTransferSides[i] = tt[sides[i]]
			cachedState = cachedState.withProperty(BlockSided.sideFromId(i), sides[i])
		}
		serverUpdate = true
	}

	override fun writeToNBT(nbt: NBTTagCompound): NBTTagCompound {
		super.writeToNBT(nbt)
		val sides = IntArray(6)
		for (i in 0..5) {
			sides[i] = allTransferSides[i].ord.toInt()
		}
		nbt.setIntArray("sides", sides)
		nbt.setInteger("module", moduleId)
		return nbt
	}

	override fun shouldRefresh(w: World, pos: BlockPos, obs: IBlockState, nbs: IBlockState): Boolean {
		return obs.block !== nbs.block || obs.getValue<BlockEnergyMachine.Variant>(BlockEnergyMachine.VARIANT) !== nbs.getValue<BlockEnergyMachine.Variant>(BlockEnergyMachine.VARIANT)
	}

	override fun getUpdatePacket(): SPacketUpdateTileEntity {
		return SPacketUpdateTileEntity(pos, blockMetadata, writeToNBT(NBTTagCompound()))
	}

	override fun onDataPacket(net: NetworkManager, pkt: SPacketUpdateTileEntity) {
		readFromNBT(pkt.nbtCompound)
	}

	override fun hasFastRenderer() = true

	fun switchSideTransfer(f: EnumFacing) {
		val s = f.index
		val v = (allTransferSides[s].ord + 1) % 3
		allTransferSides[s] = TransferType.values()[v]
		cachedState = cachedState.withProperty(BlockSided.sideFromId(s), v)
		MCFluxNetwork.toDimension(Msg.update(pos, allTransferSides), world.provider.dimension)
		markDirty()
	}

	fun getTransferSide(f: EnumFacing) = sideValues[f.index]

	fun updateTransferSides(tts: Array<TransferType>) {
		for (i in 0..5) {
			allTransferSides[i] = tts[i]
			cachedState = cachedState.withProperty(BlockSided.sideFromId(i), tts[i].ord.toInt())
		}
	}

	private fun moduleEnergyDistributor(i: Int, m: Int): Int {
		var i = i
		while (i < m) {
			val tt = allTransferSides[i]
			if (tt === TransferType.NONE) {
				i++
				continue
			}
			var f = EnumFacing.VALUES[i]
			val te = world.getTileEntity(pos.offset(f, 1))
			if (te == null) {
				i++
				continue
			}
			f = f.opposite
			val ea = FLU.getEnergySafely(te, f)
			if (ea == null) {
				i++
				continue
			}
			when (tt) {
				TransferType.INPUT -> sideValues[i] = ea.to(bat!!, (MCFluxConfig.ENERGY_DIST_TRANS * 2).toLong()) / 2
				TransferType.OUTPUT -> sideValues[i] = bat!!.to(ea, (MCFluxConfig.ENERGY_DIST_TRANS * 2).toLong()) / 2
				else -> sideValues[i] = 0
			}
			i++
		}
		return 0
	}

	private fun moduleChunkCharger(i: Int, m: Int): Int {
		var i = i
		while (i < m) {
			val tt = allTransferSides[i]
			if (tt === TransferType.NONE) {
				i++
				continue
			}
			val f = EnumFacing.VALUES[i]
			val bpc = pos.offset(f, 16)
			val ebc = wce!!.getEnergyChunk(bpc.x, bpc.y, bpc.z)
			when (tt) {
				TransferType.INPUT -> sideValues[i] = ebc.to(bat!!, (MCFluxConfig.CHUNK_CHARGER_TRANS * 2).toLong()) / 2
				TransferType.OUTPUT -> sideValues[i] = bat!!.to(ebc, (MCFluxConfig.CHUNK_CHARGER_TRANS * 2).toLong()) / 2
				else -> sideValues[i] = 0
			}
			i++
		}
		return 0
	}
}
